### PR TITLE
xformers attention with packing

### PR DESCRIPTION
**Pull request by winglian**
# Description
This PR adds packing support when using xformers attention for training. xformers supports fp16 attention, allowing us to train on T4 devices. We have to patch over FA2, as there are some attention checks that handle the attention masking that aren't easily overridable as we would have to do it for every model arch.

### CODE REVIEW

File: src/axolotl/utils/callbacks/__init__.py [hunk 1]
```diff
@@ -868,3 +868,29 @@ def on_epoch_end(
...
+            self, args, state, control, **kwargs
+        ):  # pylint: disable=unused-argument
+            """
+            handle T4 gpu, we need to convert attention to eager for inference
+            """
+            if "Tesla T4" in self.gpu_name and self.cfg.xformers_attention:
+                trainer.model.eval()
+            trainer.model.config._attn_implementation = (  # pylint: disable=protected-access
+                "eager"
+            )
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/__init__.py]: Do you mean to indent this block?

File: src/axolotl/utils/config/__init__.py [hunk 2]
```diff
@@ -70,6 +70,9 @@ def resolve_dtype(cfg):
             if cfg.fp16 is None and not cfg.float16:
                 cfg.fp16 = True
 
+    if cfg.fp16 and cfg.bf16 == "auto":
+        cfg.bf16 = False
+
```
Comment (NanoCode012) [inline src/axolotl/utils/config/__init__.py:75]: Would be good to note this behavior of `auto` in documentation, `config.qmd`.

There may be unexpected consequences for users if they have both enabled, but can use bf16.

File: src/axolotl/monkeypatch/attention/__init__.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    from .xformers import xformers_attention_forward
+
+    ALL_ATTENTION_FUNCTIONS["flash_attention_2"] = xformers_attention_forward
+
+
+def unpatch_xformers_attn_over_fa2():
+    from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS
+
+    ALL_ATTENTION_FUNCTIONS["flash_attention_2"] = flash_attention_forward()
```
Comment (NanoCode012) [inline src/axolotl/monkeypatch/attention/__init__.py:19]: I'm curious why we don't add a new attention implementation in the monkeypatch itself?

Comment (winglian) [reply src/axolotl/monkeypatch/attention/__init__.py:19]: we need to set it to use flash attention because there are other places the modeling code makes decisions based off of this.
Screenshot 2025-05-06 at 8 54 37 AM

File: src/axolotl/utils/config/__init__.py (hunk 2 above)
Comment (winglian) [reply src/axolotl/utils/config/__init__.py:75]: docs updated
